boto3/dynamodb/types.py

```diff
@@ -0,0 +1,297 @@
+            objects are used to be able to round-trip the Python type.
+
+        :rtype: dict
+        :returns: A dictionary that represents a dynamoDB data type. These
+            dictionaries can be directly passed to botocore methods.
+        """
+        dynamodb_type = self._get_dynamodb_type(value)
+        serializer = getattr(self, '_serialize_%s' % dynamodb_type.lower())
+        return {dynamodb_type: serializer(value)}
+
+    def _get_dynamodb_type(self, value):
+        dynamodb_type = None
+
+        if self._is_null(value):
+            dynamodb_type = NULL
+
+        elif self._is_boolean(value):
+            dynamodb_type = BOOLEAN
+
+        elif self._is_number(value):
+            dynamodb_type = NUMBER
+
+        elif self._is_string(value):
+            dynamodb_type = STRING
+
+        elif self._is_binary(value):
+            dynamodb_type = BINARY
+
+        elif self._is_type_set(value, self._is_number):
+            dynamodb_type = NUMBER_SET
+
+        elif self._is_type_set(value, self._is_string):
+            dynamodb_type = STRING_SET
+
+        elif self._is_type_set(value, self._is_binary):
+            dynamodb_type = BINARY_SET
+
+        elif self._is_map(value):
+            dynamodb_type = MAP
+
+        elif self._is_list(value):
+            dynamodb_type = LIST
+
+        else:
+            msg = 'Unsupported type "%s" for value "%s"' % (type(value), value)
+            raise TypeError(msg)
+
+        return dynamodb_type
+
+    def _is_null(self, value):
+        if value is None:
+            return True
+        return False
+
+    def _is_boolean(self, value):
+        if isinstance(value, bool):
+            return True
+        return False
+
+    def _is_number(self, value):
+        if isinstance(value, (six.integer_types, Decimal)):
+            return True
+        elif isinstance(value, float):
+            raise TypeError(
+                'Float types are not supported. Use Decimal types instead.')
+        return False
+
+    def _is_string(self, value):
+        if isinstance(value, six.string_types):
+            return True
+        return False
+
+    def _is_binary(self, value):
+        if isinstance(value, Binary):
+            return True
+        elif isinstance(value, bytearray):
+            return True
+        elif six.PY3 and isinstance(value, six.binary_type):
+            return True
+        return False
+
+    def _is_set(self, value):
+        if isinstance(value, Set):
+            return True
+        return False
+
+    def _is_type_set(self, value, type_validator):
+        if self._is_set(value):
+            if False not in map(type_validator, value):
+                return True
+        return False
+
+    def _is_map(self, value):
+        if isinstance(value, Mapping):
+            return True
+        return False
+
+    def _is_list(self, value):
+        if isinstance(value, list):
```

Comment:
@jamesls 
I do not think we can use Sequence because it includes strings, unicode, etc:
https://docs.python.org/2/library/stdtypes.html#sequence-types-str-unicode-list-tuple-bytearray-buffer-xrange